File: Program1/Ruby/src/FixedPointNumber.rb
class FixedPointNumber
  attr_reader :int_val, :q_val

  def initialize(val, q)
    super()
    @q_val = q
    if val.is_a?(Integer)
      @int_val = val
    else
      @int_val = (val * 2.0 ** q_val).to_i
    end

  end

  def to_double
    return (int_val.to_f / 2.0 ** q_val).to_f
  end

  def to_q_val(q)
    q_diff = q - @q_val
    if(q_diff > 0)
      new_int_val = @int_val << q_diff.abs()
    else
      new_int_val = @int_val >> q_diff.abs()
    end

    return FixedPointNumber.new(new_int_val, q_diff)
  end


  def to_s
    formatted_double = "%.6f" % self.to_double
    return "#{@int_val}, #{@q_val}: #{formatted_double}"
  end

  def ==(other_number)
    return @int_val == other_number.int_val && @q_val == other_number.q_val
  end

  def plus(other_number, q_val)
    converted_this = to_q_val(q_val)
    converted_other = other_number.to_q_val(q_val)
    return FixedPointNumber.new(converted_other.int_val + converted_this.int_val, q_val)
  end

end

File: Program1/Java/src/FixedPointNumber.java
public class FixedPointNumber {
    private int intVal;
    private int qVal;

    public FixedPointNumber(double x, int q){
        double dval = (x *  Math.pow(2,q));
        intVal = (int)(dval);
        this.qVal = q;
    }

    public FixedPointNumber(int x, int q){
        this.intVal = x;
        this.qVal = q;
    }

    public double toDouble(){
        return (double) this.intVal / Math.pow(2, this.qVal);

    }

    public FixedPointNumber toQVal(int q){
        int qDiff = q - this.qVal;
        int newIntVal;
        if(qDiff > 0){
            newIntVal = this.intVal << Math.abs(qDiff);
        }
        else {
            newIntVal = this.intVal >> Math.abs(qDiff);
        }

        return new FixedPointNumber(newIntVal,q);
    }

    @Override
    public String toString(){
        return String.format("%d, %d: %.6f",intVal, qVal, toDouble());
    }

    public boolean equals(Object p){
        if(p.getClass() != this.getClass()){
            return false;
        }
        FixedPointNumber num = (FixedPointNumber )p;
        return this.intVal == num.intVal && this.qVal == num.qVal;
    }

    public FixedPointNumber plus(FixedPointNumber p, int resultQ){
        FixedPointNumber thisConverted = this.toQVal(resultQ);
        FixedPointNumber otherConverted = p.toQVal(resultQ);

        return new FixedPointNumber(thisConverted.intVal + otherConverted.intVal, resultQ);

    }
}
